### PR TITLE
Add delay option to response

### PR DESCRIPTION
**Pull request by Darkclainer**
Now you can specify delay with "delay" field inside "response" section
(inside imposter). Delay value should be string that can be parsed by
time.ParseDuration or two such string concatenated by ":". For example
"1s" or "5s:10s". If you specified range, actual latency can be any
value within specified range.

Issue #32.

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
Found some fixes!

_P.S. share your ideas, feedbacks or issues with us at https://github.com/fixmie/feedback (this message will be removed after the beta stage)._

**Review by aperezg — APPROVED** [review]
For me is fine, great job! what do you think @joanlopez ?

**Review by aperezg — CHANGES_REQUESTED** [review]
You would need to resolve the conflicts with the master branche

**Comment by Darkclainer** [comment]
I hope I didn't mess up.

File: internal/server/http/handler.go [hunk 1]
```diff
@@ -5,11 +5,13 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"time"
 )
 
 // ImposterHandler create specific handler for the received imposter
 func ImposterHandler(imposter Imposter) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(imposter.Response.Delay.Delay())
```
Comment (joanlopez) [inline internal/server/http/handler.go]: I'd try to respect the Law of Demeter (LoD), avoiding multiple "dot" calls (A.B.C.D), because it makes the code highly coupled and it can be problematic on future changes.

File: internal/server/http/response_delay.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	var offset int64
+	if len(inputParts) == 2 {
+		maxDelay, err := time.ParseDuration(inputParts[1])
+		if err != nil {
+			return err
+		}
+		offset = int64(maxDelay) - int64(minDelay)
+	}
+	if offset < 0 {
+		return errors.New("second time point is before first time point")
```
Comment (joanlopez) [inline internal/server/http/response_delay.go]: IMHO, the error message here seems to be a bit confusing. I'd tell directly that one value is lower/greater than the other (so, it's an invalid input). I'd avoid to refer those values as they are real (date)times (where lower/greater really means before/after).

File: internal/server/http/response_delay_test.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+import (
+	"encoding/json"
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+)
+
+func TestResponceDelayUnmarshal(t *testing.T) {
```
Comment (joanlopez) [inline internal/server/http/response_delay_test.go]:
```suggestion
func TestResponseDelayUnmarshal(t *testing.T) {
```

File: internal/server/http/response_delay_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			input: `"5s:1"`,
+			err:   errors.New("error"),
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			var delay ResponseDelay
+			err := json.Unmarshal([]byte(tc.input), &delay)
+			if err != nil && tc.err == nil {
+				t.Fatalf("not expected any erros and got: %v", err)
```
Comment (joanlopez) [inline internal/server/http/response_delay_test.go]:
```suggestion
				t.Fatalf("not expected any error and got: %v", err)
```

File: internal/server/http/response_delay_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+			}
+			if !reflect.DeepEqual(tc.delay, delay) {
+				t.Fatalf("expected: %v, got: %v", tc.delay, delay)
+			}
+
+		})
+	}
+}
+
+func TestResponceDelayDelay(t *testing.T) {
```
Comment (joanlopez) [inline internal/server/http/response_delay_test.go]:
```suggestion
func TestResponseDelay(t *testing.T) {
```

**Review by joanlopez — CHANGES_REQUESTED** [review]
In general it looks really good, I left just few comments. 
Thank you so much for contributing! 👍

File: internal/server/http/handler.go (hunk 1 above)
Comment (Darkclainer) [reply internal/server/http/handler.go]: The most obvious solution, that I can imagine is to add method Delay, to Response struct.

File: internal/server/http/response_delay.go (hunk 2 above)
Comment (Darkclainer) [reply internal/server/http/response_delay.go]: I agree.
```suggestion
		return errors.New("second value should be greater than the first one")
```

**Review by ghost — COMMENTED** [review]
Found some fixes!

_P.S. share your ideas, feedbacks or issues with us at https://github.com/fixmie/feedback (this message will be removed after the beta stage)._

**Comment by Darkclainer** [comment]
Thank you for review. I've made corresponding changes to code. Should I squash all commits to a single one?

**Comment by aperezg** [comment]
> Thank you for review. I've made corresponding changes to code. Should I squash all commits to a single one?

No worry about that, when we will merge into master, we can do the squash :)

File: internal/server/http/handler.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 // ImposterHandler create specific handler for the received imposter
 func ImposterHandler(imposter Imposter) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		time.Sleep(imposter.Response.Delay.Delay())
+		time.Sleep(imposter.Response.GetDelay())
```
Comment (joanlopez) [inline internal/server/http/handler.go]: Nitpick: just `Delay()` seems even better for me, but yeah, that was the idea.

Thanks 😇

File: internal/server/http/imposter.go [hunk 7]
```diff
@@ -37,6 +38,11 @@ type Response struct {
 	Body     string             `json:"body"`
 	BodyFile *string            `json:"bodyFile"`
 	Headers  *map[string]string `json:"headers"`
+	Delay    ResponseDelay      `json:"delay"`
+}
+
+func (r *Response) GetDelay() time.Duration {
```
[the remark this answers is not in the document]
Comment (joanlopez) [reply internal/server/http/imposter.go]: Follow the `fixmie` advice and add the docs properly. 

IMHO, here's where you can add some hints about what happens under the hood.

File: internal/server/http/response_delay.go [hunk 8]
```diff
@@ -52,7 +52,7 @@ func (d *ResponseDelay) UnmarshalJSON(data []byte) error {
 		offset = int64(maxDelay) - int64(minDelay)
 	}
 	if offset < 0 {
-		return errors.New("second time point is before first time point")
+		return errors.New("second value should be greater than the first one")
```
Comment (joanlopez) [inline internal/server/http/response_delay.go]: 👏

**Review by joanlopez — CHANGES_REQUESTED** [review]

**Comment by Darkclainer** [comment]
OK, thank you for advices!

File: internal/server/http/imposter.go [hunk 9]
```diff
@@ -17,6 +18,11 @@ type Imposter struct {
 	Response Response `json:"response"`
 }
 
+// GetDelay returns delay for response that user can specify in imposter config
+func (i *Imposter) GetDelay() time.Duration {
```
Comment (aperezg) [inline internal/server/http/imposter.go]:
```suggestion
func (i *Imposter) Delay() time.Duration {
```

File: internal/server/http/handler.go [hunk 10]
```diff
@@ -5,11 +5,13 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"time"
 )
 
 // ImposterHandler create specific handler for the received imposter
 func ImposterHandler(imposter Imposter) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(imposter.GetDelay())
```
Comment (aperezg) [inline internal/server/http/handler.go]:
```suggestion
if imposter.Delay() > 0 {
  time.Sleep(imposter.Delay()
}
```

Comment (aperezg) [reply internal/server/http/handler.go]: With that avoid to call sleep even when time is 0

**Review by aperezg — CHANGES_REQUESTED** [review]
For me look good, check the latest change that I suggest, and add an example that how it's works in the README.md an thats all :) good job

File: internal/server/http/handler.go (hunk 10 above)
Comment (Darkclainer) [reply internal/server/http/handler.go]: I'm not against it.,just thought it doesn't make a lot difference. Thanks!

**Review by ghost — COMMENTED** [review]
Found some fixes!

_P.S. share your ideas, feedbacks or issues with us at https://github.com/fixmie/feedback (this message will be removed after the beta stage)._

**Comment by Darkclainer** [comment]
I have added documentation to readme but only for delay, to keep this commit small. Not sure about structure.

File: README.md [hunk 11]
```diff
@@ -1,8 +1,3 @@
-[![CircleCI](https://circleci.com/gh/friendsofgo/killgrave/tree/master.svg?style=svg)](https://circleci.com/gh/friendsofgo/killgrave/tree/master)
-[![Version](https://img.shields.io/github/release/friendsofgo/killgrave.svg?style=flat-square)](https://github.com/friendsofgo/killgrave/releases/latest)
-[![codecov](https://codecov.io/gh/friendsofgo/killgrave/branch/master/graph/badge.svg)](https://codecov.io/gh/friendsofgo/killgrave)
-[![Go Report Card](https://goreportcard.com/badge/github.com/friendsofgo/killgrave)](https://goreportcard.com/report/github.com/friendsofgo/killgrave)
-[![FriendsOfGo](https://img.shields.io/badge/powered%20by-Friends%20of%20Go-73D7E2.svg)](https://friendsofgo.tech)
```
Comment (aperezg) [inline README.md]: Why did you delete the badges?

File: README.md [hunk 12]
```diff
@@ -242,6 +237,8 @@ curl --header "Content-Type: application/json" \
   http://localhost:3000/gophers
 ```
 
+
+
```
Comment (aperezg) [inline README.md]: unnecessary break lines

File: README.md [hunk 13]
```diff
@@ -273,6 +270,21 @@ In the CORS section of the file you can find the next options:
...
   Represent the **Access-Control-Allow-Credentials header** you must indicate if true or false
+  
+## Imposter configuration file 
+Imposter file consists of JSON array of imposters. Each imposter can have two sections: `request` that describes request and `response`.
+
+### Request section
+
+### Response section
+Response section describes response that imposter server will send on matching request. Response can have next fields:
+
```
Comment (aperezg) [inline README.md]: I don't understand this section, I think is not relevant at all.

File: README.md [hunk 14]
```diff
@@ -273,6 +270,21 @@ In the CORS section of the file you can find the next options:
...
+#### delay
+Time that server waits before response. This can help simulate network issues, or server high load. You must write delay as string with postfix indicating time unit (see [this](https://golang.org/pkg/time/#ParseDuration) for more info about actual format). Also you can specify minimum and maximum delays using separator ':', server respond delay will be choosen at random between this values.
+```
+"delay": "1s:3m"
+```
+Default value is no delay at all.
```
Comment (aperezg) [inline README.md]: There are an example on the section ##Using Killgrave, with a yaml file, you would need to add in this example your delay and before of the paragraph `The option cors still...` add your text:
<pre>
The `delay` option is a time that server waits before response. This can help simulate network issues, or server high load. You must write delay as string with postfix indicating time unit (see [this](https://golang.org/pkg/time/#ParseDuration) for more info about actual format). Also you can specify minimum and maximum delays using separator ':', server respond delay will be choosen at random between this values.

Default value is no delay at all.
<pre>

**Review by aperezg — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (Darkclainer) [reply README.md]: Accidentally, I have just tested speed of generating preview, and forgot to revert this changes...

File: README.md (hunk 13 above)
Comment (Darkclainer) [reply README.md]: I've thought documentation need full description of imposter files anyway.

Comment (aperezg) [reply README.md]: Yes, you're right we're working on it, and we want to create an entire page to explain all application, and how to use each part, but for now, in we want to keep the README clear. So for now, follow my advice to try to merge your branch

**Comment by Darkclainer** [comment]
Sorry for delay, I was busy IRL. 
Honestly, I didn't quite understand why option for imposter file should be explained in section related to YAML config, but maybe it's better this way.

**Review by aperezg — APPROVED** [review]

**Comment by aperezg** [comment]
Great job dude! :D
